File: tests/deepflow-python/app/services/user_service.py
"""User service - HOP 3: Business logic layer.

Receives tainted data from routes and passes to processors.
"""

from app.processors.data_transformer import DataTransformer


class UserService:
    """User business logic service.

    HOP 3: Receives tainted query from routes, passes to processors.
    No sanitization occurs at this layer.
    """

    def __init__(self):
        self.transformer = DataTransformer()

    async def search(self, query: str) -> dict:
        """Search for users by query string.

        HOP 3: Service layer passes tainted query to processor.

        Args:
            query: TAINTED user input from request.query_params

        Returns:
            Search results from database
        """
        # Pass tainted query to transformer (HOP 4)
        transformed = self.transformer.prepare_search(query)
        return transformed

    async def get_by_id(self, user_id: str) -> dict:
        """Get user by ID.

        Args:
            user_id: TAINTED user input from path parameter
        """
        return self.transformer.prepare_lookup(user_id)

File: tests/deepflow-typescript/src/services/user.service.ts
/**
 * User service - HOP 3: Business logic layer.
 *
 * Receives tainted data from controllers and passes to processors.
 */

import { DataTransformer } from '../processors/data.transformer';

export class UserService {
  private transformer: DataTransformer;

  constructor() {
    this.transformer = new DataTransformer();
  }

  /**
   * Search for users by query string.
   *
   * HOP 3: Service layer passes tainted query to processor.
   *
   * @param query - TAINTED user input from request.query
   */
  async search(query: string): Promise<any> {
    // Pass tainted query to transformer (HOP 4)
    return this.transformer.prepareSearch(query);
  }

  /**
   * Get user by ID.
   *
   * @param userId - TAINTED user input from path parameter
   */
  async getById(userId: string): Promise<any> {
    return this.transformer.prepareLookup(userId);
  }

  /**
   * Update user settings.
   *
   * @param userId - User ID
   * @param settings - TAINTED settings object - Prototype Pollution vector
   */
  async updateSettings(userId: string, settings: any): Promise<any> {
    return this.transformer.prepareSettingsUpdate(userId, settings);
  }

  /**
   * Filter users by criteria.
   *
   * @param filter - TAINTED filter object - NoSQL injection vector
   */
  async filter(filter: any): Promise<any> {
    return this.transformer.prepareFilter(filter);
  }
}
